### DISCUSSION TITLE
skywalking for odoo

### DISCUSSION THREAD

**Discussion by ZKANGGIT** [apache/skywalking#12262] [General]
I have recently started working on integrating Skywalking with Odoo again. I have already developed two plugins and one decorator based on Odoo's framework, which work perfectly in single-process mode. However, when Odoo is configured with multiple workers, monitoring cannot be achieved. Can you help me solve this problem?

**Comment by wu-sheng** [comment]
Do you follow the plugin dev doc? Context needs to be propagated.

**Reply by ZKANGGIT** [reply]
开发文档看了，主要照着其他的插件开发的，下面是我的源码：
启动文件：
__import__('os').environ['TZ'] = 'UTC'
import odoo
from skywalking import agent, config

config.init()
config.agent_collector_backend_services = "10.11.65.222:12800"
config.agent_name = "localhost"
config.agent_protocol = "http"
config.agent_log_reporter_active = True
config.agent_log_reporter_level = "DEBUG"
agent.start()


if __name__ == "__main__":
    odoo.cli.main()

插件1：
# sw_odoo_json.py
from skywalking.trace.carrier import Carrier
from skywalking import Layer, Component, config
from skywalking.trace.context import get_context, NoopContext
from skywalking.trace.span import NoopSpan,EntrySpan
from skywalking.trace.tags import TagHttpMethod, TagHttpStatusCode, TagHttpURL, TagHttpParams

# 插件支持的版本信息
support_matrix = {
    'odoo': {
        '14': ['14.0'],
    }
}

def install():
    """
    安装插件，注册对 Odoo 核心方法的监控。
    """
    from odoo.http import JsonRequest

    original_dispatch = JsonRequest.dispatch

    def _sw_dispatch(self):

        headers = self.httprequest.headers
        path = self.httprequest.path
        carrier = Carrier()
        for item in carrier:
            if item.key.capitalize() in self.httprequest.headers:
                item.val = headers[item.key.capitalize()]
        span = NoopSpan(NoopContext()) if config.ignore_http_method_check(path) \
            else get_context().new_entry_span(op=path.split('?')[0], carrier=carrier)

        with span:
            url = self.httprequest.url
            span.layer = Layer.Http
            span.component = Component.Requests
            span.peer = self.httprequest.host
            span.tag(TagHttpMethod(self.httprequest.method))
            span.tag(TagHttpURL(url))
            args = self.httprequest.data
            span.tag(TagHttpParams(str(args)))

            response = original_dispatch(self)

            span.tag(TagHttpStatusCode(response.status_code))

            if response.status_code >= 400:
                span.error_occurred = True

            return response

    JsonRequest.dispatch = _sw_dispatch

if __name__ == "__main__":
    install()


插件2：
# sw_odoo_http.py
from skywalking.trace.carrier import Carrier
from skywalking import Layer, Component, config
from skywalking.trace.context import get_context, NoopContext
from skywalking.trace.span import NoopSpan
from skywalking.trace.tags import TagHttpMethod, TagHttpStatusCode, TagHttpURL, TagHttpParams

# 插件支持的版本信息
support_matrix = {
    'odoo': {
        '14': ['14.0'],
    }
}

def install():
    """
    安装插件，注册对 Odoo 核心方法的监控。
    """
    from odoo.http import HttpRequest

    original_dispatch = HttpRequest.dispatch

    def _sw_dispatch(self):
        headers = self.httprequest.headers
        path = self.httprequest.path
        carrier = Carrier()
        for item in carrier:
            if item.key.capitalize() in self.httprequest.headers:
                item.val = headers[item.key.capitalize()]
        span = NoopSpan(NoopContext()) if config.ignore_http_method_check(path) \
            else get_context().new_entry_span(op=path.split('?')[0], carrier=carrier)

        with span:
            url = self.httprequest.url
            span.layer = Layer.Http
            span.component = Component.Requests
            span.peer = self.httprequest.host
            span.tag(TagHttpMethod(self.httprequest.method))
            span.tag(TagHttpURL(url))
            args = [item for item in self.httprequest.args.items()]
            args += [item for item in self.httprequest.form.items()]
            span.tag(TagHttpParams(str(args)))

            response = original_dispatch(self)

            span.tag(TagHttpStatusCode(response.status_code))

            if response.status_code >= 400:
                span.error_occurred = True

            return response

    HttpRequest.dispatch = _sw_dispatch


监控odoo model的装饰器：
def fun_wrapper(func):
    def wrapper(*args, **kwargs):
        context = get_context()
        span = context.new_local_span(op=func.__name__)
        span.layer = Layer.Unknown
        span.component = Component.Unknown
        with span:
            span.peer = func.__module__
            param = {
                "args": args,
                "kwargs": kwargs
            }
            span.tag(TagHttpParams(str(param)))
            return func(*args, **kwargs)
    return wrapper

def trace_class(cls):
    for name, method in vars(cls).items():
        if callable(method):
            setattr(cls, name, fun_wrapper(method))
    return cls

**Reply by ZKANGGIT** [reply]
![QQ截图20240529111712](https://github.com/apache/skywalking/assets/36831988/dfadcb25-1996-4d8b-9051-e09d02c3af8d)
这个是odoo单进程模式下的效果

**Reply by wu-sheng** [reply]
Please use English. Use slack if you want to talk in Chinese.

**Reply by ZKANGGIT** [reply]
> Please use English. Use slack if you want to talk in Chinese.
ok，I would like to confirm something. I found information that context propagation in SkyWalking is implemented through ThreadLocal, which means that cross-process context sharing is not achievable in SkyWalking. From the results of my current usage, it seems that the agent indeed does not achieve cross-process monitoring.

**Reply by wu-sheng** [reply]
Your plugin takes responsibility to propogate that. You could refer to a RPC framework plugin.

**Reply by ZKANGGIT** [reply]
Okay, thank you. I'll take a look at the relevant code.

**Reply by ZKANGGIT** [reply]
![image](https://github.com/apache/skywalking/assets/36831988/6fcf1289-66b1-4a0c-81a8-1eb115d1a29c)
Is this it?

**Comment by ZKANGGIT** [comment]
<img width="1709" alt="企业微信截图_17169532421097" src="https://github.com/apache/skywalking/assets/36831988/8898c2fe-ff11-4c65-b8cb-88fda1ebe551">
